pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,493 @@
+			entry := m.latencyStore.GetNodeIPLatencyEntryByKey(destIP)
+			if entry == nil {
+				klog.Warning("Failed to get node entry by destIP", "destIP", destIP)
+				continue
+			}
+
+			// Parse the ICMP message
+			var msg *icmp.Message
+			if isIPv4 {
+				msg, err = icmp.ParseMessage(ProtocolICMP, readBuffer[:n])
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+				if msg.Type != ipv4.ICMPTypeEchoReply {
+					klog.Warning("Failed to match ICMPTypeEchoReply")
+					continue
+				}
+			} else {
+				msg, err = icmp.ParseMessage(ProtocolIPv6ICMP, readBuffer)
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+				if msg.Type != ipv6.ICMPTypeEchoReply {
+					klog.Warning("Failed to match ICMPTypeEchoReply")
+					continue
+				}
+			}
+
+			echo, ok := msg.Body.(*icmp.Echo)
+			if !ok {
+				klog.Warning(nil, "Failed to assert type as *icmp.Echo")
+				continue
+			}
+
+			klog.InfoS("Recv ICMP message", "IP", destIP, "SeqID", entry.SeqID, "echo", echo)
+
+			// Parse the ICMP data
+			if entry.SeqID != uint32(echo.Seq) {
+				klog.Warning("Failed to match seqID", "entry.SeqID", entry.SeqID, "echo.Seq", echo.Seq)
+				continue
+			}
+
+			// Calculate the round-trip time
+			end := time.Now()
+			rtt := end.Sub(entry.LastSendTime)
+
+			// Update the latency store
+			m.latencyStore.UpdateNodeIPLatencyEntryByKey(destIP, &NodeIPLatencyEntry{
+				SeqID:           entry.SeqID,
+				LastSendTime:    entry.LastSendTime,
+				LastRecvTime:    end,
+				LastMeasuredRTT: rtt,
+			})
+		}
+	}
+}
+
+func (m *NodeLatencyMonitor) pingAll(ipv4Socket, ipv6Socket net.PacketConn) {
+	// Get all node internal/external IP.
+	nodeIPs := m.latencyStore.ListNodeIPs()
+	klog.InfoS("Start to ping all nodes")
+	for name, toIPs := range nodeIPs {
+		for _, toIP := range toIPs {
+			klog.InfoS("Start to ping node", "Node name", name, "Node IP", toIP)
+			if toIP.To4() != nil && ipv4Socket != nil {
+				if err := m.sendPing(ipv4Socket, toIP); err != nil {
+					klog.InfoS("Failed to send ICMP message to node", "Node name", name, "Node IP", toIP)
+				}
+			} else if toIP.To16() != nil && ipv6Socket != nil {
+				if err := m.sendPing(ipv6Socket, toIP); err != nil {
+					klog.InfoS("Failed to send ICMP message to node", "Node name", name, "Node IP", toIP)
+				}
+			} else {
+				klog.InfoS("Failed to send ICMP message to node", "Node name", name, "Node IP", toIP)
+			}
+		}
+	}
+}
+
+func (m *NodeLatencyMonitor) testPrint() {
+	// Print all connection status for debug
+	// It will be removed when collector is ready
+	klog.InfoS("Finish to ping all nodes")
+	entries := m.latencyStore.ListLatencies()
+	for key, entry := range entries {
+		klog.InfoS("NodeIPLatency status", "Key", key, "Entry", entry)
+	}
+}
+
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	// Top level goroutine to handle termination
+	go func() {
+		<-stopCh
+	}()
```

Comment:
this doesn't do anything actually
typically we have `<-stopCh` at the end of the `Run` function

Reply:
Ok, I have fixed it.